### DISCUSSION TITLE
Export backups daily

### DISCUSSION THREAD

**Discussion by licaon-kter** [iSoron/uhabits#178] [Feature Requests] [answered]
Since #26 got closed, I still wish for this though, not sure why a just straight forward "export full backup daily" can't be implemented at least.

I use Syncthing but as a means for backup only, not for external editing.

**Comment by sciamano** [comment]
you could achieve that by using https://github.com/jensstein/oandbackup for daily backup and then get files out ouf the device by Syncthing

**Comment by licaon-kter** [comment]
Why would I bother with that if I have root?

No, root is a precious rare thing these days, I don't have root, and since the app can already save files, I wish for a daily file save in `/sdcard/Android/data/appname/files` folder, yeah the same folder the app opens up when asked to import a backup.

**Comment by iSoron** [comment]
I am okay with exporting a daily backup to the SD card. In fact, I think it could be even enabled by default. Most of the code for this feature is already there (see [ListHabitsController.java](https://github.com/iSoron/uhabits/blob/0a5677211e041a377d9c1bdc057da0fd5a27d9fc/app/src/main/java/org/isoron/uhabits/activities/habits/list/ListHabitsController.java#L111-L117)), but in addition to saving the file externally, we would also need to (1) automatically delete old backup files, and (2) add an option to disable these daily backups.

**Comment by licaon-kter** [comment]
Yeah, maybe add just a counter "keep last X days" or something, where 0 means disabled.

**Comment by filmerjarred** [comment]
+1 this would make integrating with services like exist.io x10 easier

**Comment by jacobmoe** [comment]
Does it make sense for this feature to include an option for a daily CSV export to Google Drive (or whatever)?

**Comment by deutrino** [comment]
I believe newish versions of Android may make exporting to various cloud storage services easier than it used to be, but I personally hope that if periodic backup does get implemented, there is an option to do so locally to the file system 🙂

**Comment by morrel** [comment]
Automatic or scheduled CSV/Spreadsheet exports would also be highly valued!

**Comment by amitkot** [comment]
I might be able to help with developing an MVP of this feature.

I was thinking along the lines of two options:
1. If this is something we thing many people would like - saving a number of daily backups (say 3) to a readable path on the sdcard.
2. If this is something only pro-users might be interested in - exposing a Tasker task for exporting the data, this way the users can decide for themselves where to save, and also trigger the action.

The second option will probably be the easiest and fastest to develop.

The first option, if developed, should maybe be kept as simple as possible. I was thinking of a simple trigger, maybe app stop, that would run a task that in turn would:
- Dump the db on external storage
- Check if there are more than 3 such dump files and delete oldest files.

Would be interested to hear (a) if option 2 would be considered and (b) what do you think about the design suggested for option 1.

**Comment by sciamano** [comment]
Hi, just my personal opinion: i think the apps should not reinvent the wheel again and again :)
so in that case i'm for option 2 and against 1 (that way it will also keep the required permission low)
And a backup really make sense only if users managed to get it out of the phone e.g. into some cloud.
Then the cloud will take care about keeping some previous versiont etc. and you do not have deal with 3 dump files (and it may happen that the number 3 does not fit all the users...)

**Comment by Clarkeyyy** [comment]
Hey, what's the status on this?

**Comment by gingerling** [comment]
Hi, I would really appreciate this feature, or a paid service from the app devs. I rely so heavily on this app now, it's been a big gsme changer for me managing my health conditions - since I started using it with widgets, it's been just amazing. But, I'm really worried about the backup situation (and my phone isn't rooted)

**Comment by r4taw2mg** [comment]
I'd be happy, if the current backup functions (full and csv) could just be automated to do their jobs once a day. I too use syncthing to copy those to a non-mobile space. Today I have to do this manually with a high risk of forgetting.

**Comment by trmendes** [comment]
Same thing here. Using syncthing to backup my android and it would be nice to have automatic daily backups on my Habits \o/

**Comment by rick-nu** [comment]
I just lost one and a half year of data. It was my fault not for thinking of making a manual backup for before re-installing my phone, yet automated daily backups would have saved me.

It would be super awesome if this feature can make it into the app!

**Comment by licaon-kter** [comment]
Nothing new?

I'll need to create a "habit" with "save backup" soon :(

It will be 100%  ¯\\\_(ツ)_/¯

**Comment by daniilrozyev** [comment]
Hi! I really need this function. Please add it.

**Comment by filmerjarred** [comment]
I knocked together a quick and dirty branch that creates a daily export. If you want the functionality bad enough feel free to use it :)

https://github.com/BrighTide/uhabits/tree/poc-add-daily-export

**Comment by woolfog** [comment]
Having backups automated would be helpful. If it could save a backup to a folder that Syncthing is watching on a daily basis, that would be awesome. 

I am rather clumsy so at least this way my data would be safe if I were to drop my phone in water or something. 

EDIT: An alternate solution I think would be to have it export the date whenever the user takes some action, like checking/unchecking a habit.

**Comment by nsnitesh7** [comment]
@iSoron - When are we targeting to work on this and release?

I love this app. Thank you so much for working on this amazing project.

**Comment by jigneshdarji91** [comment]
Is the scope of this task to add support for daily SD card export only? Or is the scope also to add Google Drive daily export?

**Comment by zaheernathani** [comment]
Its almost 4 Years this thread is open. Any timeline on this feature to be implemented ?

**Comment by david-macleod** [comment]
Would also be really keen to see this feature 🙂

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by lerra** [comment]
@BrighTide any plans to merge it? :-)

**Comment by filmerjarred** [comment]
@lerra ha no sorry, I don't really know Java or the project. This was just a hack to get it working for me that others might be able to use 😅

**Comment by iSoron** [comment, marked as answer]
This has been fixed in dc0b8deccf40be747fc2b34d7c836e4808146c51. The app will now automatically save a backup file to the SD card whenever the main activity is launched, assuming that no backup was generated within the last 24 hours. If there is already a very recent backup (<24 hours), the app should skip the auto backup.

To avoid cluttering, the app will only keep a few files in the backup folder (currently 5 old backups + the newest one). All the other files in the backup folder should get automatically deleted. If you want to keep very old backup files for some reason, you will need to keep your own copy elsewhere.

This feature should become available in Loop 2.0.

**Reply by mshuflin** [reply]
Is there a default path for this backup? Is it the db file or csv file that is automatically stored? I can't seem to find either

**Reply by ropery** [reply]
I agree with @felixwiemuth below that this feature should be documented in the App.
I would also suggest adding a link to open the backup location in a file manager (at lease initially, before maybe introducing configurations for custom backup location & number of backups to keep later on), allowing the user to navigate there quickly & copy any .db file elsewhere they want to preserve before auto-deletion.

**Reply by steel-steven-87** [reply]
@iSoron  I'm confused by the answer. What is the location of the backup and can it be changed? I would like to choose a path which gets synchronized by a cloud provider app.

**Reply by holocronweaver** [reply]
Based on the code linked, I found the backup location at `Android/data/org.isoron.uhabits.database/files/Backups`, which is definitely not ideal for syncing as it requires root permissions to access.

**Reply by jonathan-s** [reply]
@iSoron when trying to reach that folder I get the message to protect your privacy please choose another folder. So being able to set the backup location would make things a lot easier.

**Comment by mahdi1234** [comment]
Backup seems to work fine, one great addition would be user defined folder to place it to e.g. into my already existing nextcloud sync folder, not to need additional tool to copy it over. Also automatic export to csv would be great.

**Comment by lerra** [comment]
I believe it will stop working in android 11 with the new app storage
permission if you ask the app to write outside its own folder

**Reply by licaon-kter** [reply]
For F-Droid, the app can ask for the extended storage permission, and continue to write.

**Comment by felixwiemuth** [comment]
What is the status of this feature, especially in Android 11? It seems to be right that this is not working anymore, I can't find any automatically generated backups (and the feature is not documented either).

To get it to work with the new storage framework it would probably be enough (and also best) to let the user choose a backup location.

My suggestions:
- Let user choose backup directory.
- Let the minimum time between backups be configurable; maybe some users also want a backup after every change (or when leaving the app / it gets suspended or after a certain delay after the last change), to increase safety even more.
- Let the user define the number of backups to keep (those using synchronization/backup apps that do versioning - like synchting - probably only want one file).

Most of these things are luckily only a setting for an existing variable and should thus be easy to implement.

Looking forward to a working auto backup which would close an important gap in this otherwise great app!

**Comment by licaon-kter** [comment]
`/storage/Android/data/org.isoron.uhabits/files/Backups/` is the path

**Comment by licaon-kter** [comment]
There's also the issue of app uninstall, which will remove that folder.

**Comment by RabidTunes** [comment]
I would also like to have this feature seems like a very good one. Great app by the way, it's absolutely fantastic :)

**Comment by NateEag** [comment]
I'm another delighted user who would really appreciate having the CSV auto-exported alongside the app DB.

Might a pull request adding the CSV to the auto-export be accepted?

**Reply by mshuflin** [reply]
I would love to have csv auto exported, that would be great!

**Reply by RestlessRabbits** [reply]
Hi, I also would love an auto export feature!
I for example would use it to transfer / sync that file to another system and check file change with my home automation system to mirror my stats there as well. This is still the best app I was able to find when tracking habits and I think the possibility to access the data would greatly improve the possibilites of the habit tracking with this app. :)

**Reply by NateEag** [reply]
@iSoron Just pinging you again - would a PR adding the CSV to the auto-export be considered?

**Reply by iSoron** [reply]
@NateEag Yes, a PR would be welcome. Note that we already auto-export a full backup daily. We could do something very similar for the CSV.

**Reply by kuba-orlik** [reply]
@iSoron 

> Note that we already auto-export a full backup daily

Awesome! How do I enable that? I can't find this in settings

EDIT: Ah, it's automatically enabled and stored in `/storage/Android/data/org.isoron.uhabits/files/Backups/`

**Reply by NateEag** [reply]
great, thanks @iSoron !

Handling it similarly to the existing auto-backup sounds great.

I likely won't take time to look at this soon, but I've put a note in my todo list to take a stab at it in March.

**Reply by ColdHotman** [reply]
> I've put a note in my todo list to take a stab at it in March.

That sounds great! Do you think it's possible to auto-export to a single CSV? I would absolutely love to have a single file with all the habits instead of two files per habit. That's a lot of files to manually import. 😅

**Reply by NateEag** [reply]
I obviously did not get to this in March - I've been really busy of late.

I'm hoping I will find some time over the summer to take a stab at it.

@ColdHotman I'm not sure if a single CSV will be realistic - I'll keep it in mind when I manage to take time to look at this.

**Comment by steel-steven-87** [comment]
I think the app needs this feature badly. The user should be able to select a location to automatically back up the data every day. That way the user would have the choice to use any cloud storage app to sync the folder to the cloud.

**Comment by ghost** [comment]
I was also searching for how to do this - a backup that isn't uploaded somewhere isn't really a backup. It's a locally stored database copy.

**Reply by NateEag** [reply]
FWIW, I use [SyncThing](https://syncthing.net/) to auto-sync the auto-backups to my laptop whenever the phone and laptop are on the same wireless network. I also have some janky scripts to auto-commit-and-push the updated repos after sync (a simpler but less granular approach would be a cron job that auto-commits-and-pushes daily or hourly).

That strategy works okay for any app that supports an auto-backup, and I use it with several of them.

**Reply by TheAmogh** [reply]
I don't have permissions to access the backup folder ```/storage/Android/data/org.isoron.uhabits/files/Backups/```. I thought recent Android versions don't allow read permissions to the app data directories. How is SyncThing able to access this location?

**Reply by NateEag** [reply]
@TheAmogh Sorry I missed this question for so long.

Looks like I was mistaken - I'd been syncing manual backups, not the automated ones. 

I'd fallen out of using this app for a lot of 2023, and I got the way I'd had it set up confused with how I have several other apps configured. My apologies.

**Comment by Pacman99** [comment]
The current auto backup feature isn't particularly useful because no backup app can access it (like nextcloud or syncthing). It would be great if the path where auto backups went was customizable or stored externally.

**Comment by adharsh** [comment]
Hey guys, here's an approach that'll automatically back up your data to Google Drive:
https://github.com/adharsh/uhabits

This patch simply saves the db to `Download/Loop-Habits-DB/Loop_Habits_Backup.db` whenever a backup is triggered (both automatic triggers and manual exports). I then use the app `Autosync for Google Drive` by MetaCtrl to automatically back this file to Google Drive every 2 hours (interval can also be set to 1 hour to 24 hours). 
If you’re worried about giving the autosync app full access to your Google Drive, you can always create a separate Google account just for this.

**Reply by brog-io** [reply]
> Hey guys, here's an approach that'll automatically back up your data to Google Drive:
> https://github.com/adharsh/uhabits
> 
> This patch simply saves the db to `Download/Loop-Habits-DB/Loop_Habits_Backup.db` whenever a backup is triggered (both automatic triggers and manual exports). I then use the app `Autosync for Google Drive` by MetaCtrl to automatically back this file to Google Drive every 2 hours (interval can also be set to 1 hour to 24 hours). 
> If you’re worried about giving the autosync app full access to your Google Drive, you can always create a separate Google account just for this.

How do I download this version?

**Reply by druckdev** [reply]
Would you mind creating a PR with your changes? That'd be great!

**Reply by OdyAsh** [reply]
> Would you mind creating a PR with your changes? That'd be great!

Up

**Comment by MihanEntalpo** [comment]
I've created a PR that adds automatic backups to a public folder with a selectable frequency: https://github.com/iSoron/uhabits/pull/2207

**Reply by mshuflin** [reply]
Amazing! This would completely automate my workflow for tracking! Currently I manually export the CSV files to syncthing and use node red to read them. I suppose I will need to redo that flow slightly with the SQL database, but that should work as well.

Thank you! Hope that is able to get merged, happy to help test if needed

**Reply by MihanEntalpo** [reply]
Anyway, you can clone my repo (`git clone git@github.com:MihanEntalpo/uhabits.git`), switch to branch `automatic_backups_to_public_folder`, build APK (https://github.com/iSoron/uhabits/blob/dev/docs/BUILD.md) and install it to your android phone, as I've made already for myself.

**Reply by lerra** [reply]
This is awesome, thanks for the contribution, looking forward to the merge!

**Reply by MihanEntalpo** [reply]
Made another PR in place of the mentioned one: https://github.com/iSoron/uhabits/pull/2209
Now it enabled build-in AutoBackup functionality to use public folder.

**Reply by lerra** [reply]
> Made another PR in place of the mentioned one: https://github.com/iSoron/uhabits/pull/2209
> Now it enabled build-in AutoBackup functionality to use public folder.
This is awesome!
